ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp

```diff
@@ -871,21 +844,128 @@ std::tuple<ttnn::Tensor, uint32_t, uint32_t, ttnn::Tensor, std::optional<ttnn::T
+        }
     }
+    uint32_t output_height = ((input_height - kernel_size[0] - ((kernel_size[0] - 1 ) * (dilation[0] - 1)) + 2 * padding[0]) / stride[0]) + 1;
+    uint32_t output_width = ((input_width - kernel_size[1] - ((kernel_size[0] - 1 ) * (dilation[0] - 1)) + 2 * padding[1]) / stride[1]) + 1;
+    auto [input_tensor_post_tm, parallel_config, tensor_manipulated] = shard_or_reshard_tensor_if_required(
+        device,
+        input_tensor,
+        conv_config,
+        batch_size,
+        output_height,
+        output_width,
+        in_channels,
+        out_channels,
+        kernel_size,
+        stride,
+        padding,
+        dilation,
+        weight_tensor.get_shape()[3],
+        input_width,
+        groups);
+    if (tensor_manipulated) {
+        if (conv_config.deallocate_activation) {
+            ttnn::Tensor input_tensor_ = input_tensor;  // TODO: allow in place modification of inputs to the op
+            input_tensor_.deallocate();
+            // ttnn::operations::core::deallocate(input_tensor_);
+        }
+        conv_config.deallocate_activation = true;
+    }
+    auto conv_out_memory_config = create_sharded_memory_config_from_parallel_config(
+        ttnn::Shape(std::array<uint32_t, 4>{1, 1, batch_size * output_height * output_width, tt::round_up(out_channels, 32)}),
+        parallel_config,
+        32);
+    auto opt_conv_op_parallel_config = determine_conv_op_parallel_config_from_conv_output_mem_config(
+        conv_out_memory_config, get_num_cores_nhw_from_parallel_config(parallel_config),
+        get_num_cores_channels_from_parallel_config(parallel_config));
+    auto opt_conv_op_block_config = determine_per_core_conv_block_config(
+        parallel_config,
+        opt_conv_op_parallel_config,
+        tt::round_up(in_channels, conv_config.input_channels_alignment),
+        conv_config.act_block_h_override,
+        conv_config.act_block_w_div,
+        kernel_size[0],
+        kernel_size[1],
+        conv_config.fp32_dest_acc_enabled,
+        conv_config.input_channels_alignment == 16);
+
+    // Validate weight and bias tensors
+    TT_FATAL(ttnn::is_tensor_on_device_or_multidevice(weight_tensor), "Weight tensor must be on device or multidevice");
+    TT_FATAL(weight_tensor.get_shape().rank() == 4, "Weight tensor must be 4D");
+    TT_FATAL(weight_tensor.get_shape()[0] == 1 && weight_tensor.get_shape()[1] == 1, "weight[0] != 1 and/or weight[1] != 1 indicates that the weight tensor has not been prepared for ttnn. Please use prepare_conv_weights_for_ttnn to prepare the weight tensor.");
+    if (bias_tensor.has_value()) {
+        TT_FATAL(ttnn::is_tensor_on_device_or_multidevice(bias_tensor.value()), "Bias tensor must be on device or multidevice");
+        TT_FATAL(bias_tensor->get_shape().rank() == 4, "Bias tensor must be 4D");
+        TT_FATAL(bias_tensor->get_layout() == Layout::TILE, "Bias tensor must be in TILE layout");
+    }
+
```

Comment:
also, should we assert for weights not on device or call conv2d_host_weights?

Reply:
Perhaps, but that would require us changing numeros tests too. Should probably do that when we do the full replacement in the next PR